addons/website_sale/controllers/main.py

```diff
@@ -1769,7 +1769,11 @@ def shop_payment_validate(self, sale_order_id=None, **post):
             return request.redirect('/shop')
 
         if order and not order.amount_total and not tx_sudo:
-            order.with_context(send_email=True).with_user(SUPERUSER_ID).action_confirm()
+            is_order_invoiced = order.with_context(send_email=True).with_user(SUPERUSER_ID)._validate_zero_amount_cart()
+            if is_order_invoiced:
+                # clean context and session, then redirect to the portal page
+                request.website.sale_reset()
```

Comment:
Yes but before i did this task the code did not have `request.website.sale_reset()` and since this task is to address a specific flow (when amount total is zero due to a discount/gift card/...) .I added a condition for `is_order_invoiced`. So as not to break original code flow

Reply:
I believe it's safe and cleaner to do that change.
That way, you wouldn't have to return a boolean value in your new method.